mycpp/leaky_types.cc

```diff
@@ -32,6 +160,128 @@ bool OmitChar(uint8_t ch, int what) {
   }
 }
 
+
+
+// #######################################
+
+
+Str* Str::upper() {
+  int len_ = len(this);
+  Str* result = AllocStr(len_);
+  char* buffer = result->data();
+  for (int char_index = 0; char_index < len_; ++char_index) {
+    buffer[char_index] = toupper(data_[char_index]);
+  }
+  return result;
+}
+
+Str* Str::lower() {
+  int len_ = len(this);
+  Str* result = AllocStr(len_);
+  char* buffer = result->data();
+  for (int char_index = 0; char_index < len_; ++char_index) {
+    buffer[char_index] = tolower(data_[char_index]);
+  }
+  return result;
+}
+
+// #######################################
+
+
+Str* Str::ljust(int width, Str* fillchar) {
+  assert(len(fillchar) == 1);
+
+  int len_ = len(this);
+  int num_fill = width - len_;
+  if (num_fill < 0) {
+    return this;
+  } else {
+    char* buf = static_cast<char*>(malloc(width));
+    char c = fillchar->data_[0];
+    memcpy(buf, data_, len_);
+    for (int i = len_; i < width; ++i) {
+      buf[i] = c;
+    }
+    return CopyBufferIntoNewStr(buf, width);
```

Comment:
Ditto I think all of these will end up as `AllocStr(width)` and then mutating that buffer

I'm a little scared by the malloc(width) because it used to be malloc(width + 1), but AllocStr will solve that problem because it takes string length and not buffer length